financial-templates-lib/price-feed/utils.js

```diff
@@ -0,0 +1,130 @@
+const lodash = require("lodash");
+const assert = require("assert");
+
+// Downloads blocks and caches them for certain time into the past.
+// Allows some in memory searches to go from timestamp to block number.
+exports.BlockHistory = (web3, blocks = []) => {
+  assert(web3, "requires web3");
+
+  // Check if we have downloaded a block by number
+  function has(number) {
+    return blocks.find(block => block.number === number);
+  }
+
+  // Used internally, but will insert a block into cache sorted by timestamp
+  function insert(block) {
+    const index = lodash.sortedIndexBy(blocks, block, "timestamp");
+    blocks.splice(index, 0, block);
+    return blocks;
+  }
+
+  // Get the block with the closest match to timestamp
```

Comment:
yes and actually im not even sure if that is the right behavior. I could run it so it actually does find the closest block. thoughts?

Reply:
If you're using this to get a price, you probably want the closest block _before_ the point in time, since that would be the prevailing price at that time (that's what people would see if they checked the price "right now" at that time since they don't have the benefit of future information). If you want this to be more general, I'm not sure -- you may want to allow the caller to specify or something....